Add suport for tracing

skoudmar opened the pull request:
This PR adds support for tracing by importing tracepoints used by rclcpp and creating new ones.

ROS 2 tracing is built on the [LTTng][] tracing framework.

### Motivation

Tracing ROS applications can help identify performance bottlenecks or measure and identify causes of latencies.

Available tools to analyze traces:

- [Trace compass][] with experimental plugin *ROS2*
- [`tracetools_analysis`][]
- My [Ros2TraceAnalyzer][]

### Overview

Traces of R2R application without this PR already include tracepoints
from `rcl` and `rmw`, but these are not sufficient for extracting
useful information. Therefore, this PR extends R2R in the following
areas:

- Creates the `r2r_tracing` module, which
  - Imports tracepoints from ROS package [`tracetools`][] (a part of `ros2_tracing`).
  - Defines new R2R specific tracepoints (for `Node::spin_once`).
  - Uses the `tracing` feature flag. Without it nothing is imported/linked, and tracepoints are no-ops.
- Calls tracepoint functions defined in `r2r_tracing` from `r2r` code.
- Adds methods `Node::subscribe_traced`, `Node::create_service_traced`,
  and `Timer::on_tick`. The first two are similar to `Node::subscribe`
  and `Node::create_service`, but add tracepoints around user-supplied
  callback. `Timer::on_tick` adds tracepoints to timer callback.
- Updates R2R so that ROS entities are created in a location with
  stable address.

### Stable memory address

Without this PR, ROS entities are first created on the stack, so initialization events like `ros2::rcl_publisher_init` use the stack address as their tracepoint event. Besides the address, these initialization events usually contain helpful information such as `topic` or `period`. R2R then moves the entity to the heap using `Box` or `Arc`. When a traced function, for example, `rcl_publish`, is later called on the entity, the tracepoint event references the heap memory address instead. Since the heap address differs from the original stack address, the traced function cannot be correctly matched with the corresponding initialization event.

TODO: update PR link

This PR solves the problem for publishers (cb10d284a61b22034aefba8e07ae05cf6ddfd05c), subscriptions (3f245145d11c2dd4a7ef0414d70dd3426ff1eaf4), services (e9dece1d8133b460e2fa10f5f8343942ea834a25), and clients (c926fa671fde3bfe77c9082c61aa30c0258b0061). For timers, the problem was already solved in PR #1.

### Importing tracepoints from [`ros2_tracing`][]

To add tracepoints compatible with `rclcpp` to the R2R code, we first need to import them. This is handled by dynamically loading [`tracetools`][] library from [`ros2_tracing`][]. [`tracetools`][] is a part of ROS, and is depended on by, e.g., `rcl` or `rclcpp`.

The following tracepoints are imported:
- `ros2:rclcpp_subscription_init`
- `ros2:rclcpp_timer_link_node`
- `ros2:rclcpp_subscription_callback_added`
- `ros2:rclcpp_service_callback_added`
- `ros2:rclcpp_timer_callback_added`
- `ros2:rclcpp_callback_register`
- `ros2:rclcpp_publish`
- `ros2:rclcpp_take`
- `ros2:callback_start`
- `ros2:callback_end`

Reference: [https://github.com/ros2/ros2_tracing/blob/rolling/doc/design_ros_2.md](https://github.com/ros2/ros2_tracing/blob/rolling/doc/design_ros_2.md)

### Defining new tracepoints

In `rclcpp`, sampling of ready entities is announced by tracepoint `rclcpp_executor_wait_for_work`. However, this tracepoint is tied to the C++ executor lifecycle, and R2R does not include executors. In R2R, sampling is done by `Node::spin_once`. Therefore, to be able to get information about sampling, we define new tracepoints:

- `r2r:spin_start`
- `r2r:spin_wake`
- `r2r:spin_timeout`
- `r2r:spin_end`

These are executed according to the following table:

| Spin result             | Tracepoint sequence                       |
| :---------------------- | :---------------------------------------- |
| Some entities are ready | (spin_start) -> (spin_wake) -> (spin_end) |
| Wait timeouted          | (spin_start) -> (spin_timeout)            |

The tracepoints are generated by crate [`lttng-ust-generate`][]. Its release version is old and does not support `usize`, so I have created a PR (https://github.com/sapphire-arches/lttng-ust-rs/pull/5) which fixes it, but the maintainer is not active, so [my fork](https://github.com/skoudmar/lttng-ust-rs) is used instead.

Additionally, the generated code needs to link with [`lttng-ust`][] library which is only available on Linux. This is the main reason the tracing support is feature-gated.

### `tracing` feature flag

Generation, inclusion, and linking are controlled by the `tracing` feature flag in `r2r_tracing`, which is also exported by `r2r`.

Without the feature flag `tracing`, nothing is imported/linked, and tracepoints are no-ops.


### New callback tracing functions

In order to support tracing of callback execution, new functions `Node::subscribe_traced`, `Node::create_service_traced`, and `Timer::on_tick` are added.

Example usage is provided in `r2r/examples/callback_tracing.rs`.

[`ros2_tracing`]: https://github.com/ros2/ros2_tracing
[`tracetools`]: https://index.ros.org/p/tracetools/
[`lttng-ust-generate`]: https://lib.rs/crates/lttng-ust-generate
[`lttng-ust`]: https://lttng.org/man/3/lttng-ust/v2.13/#
[LTTng]: https://lttng.org/
[Trace compass]: https://eclipse.dev/tracecompass/
[`tracetools_analysis`]: https://github.com/ros-tracing/tracetools_analysis
[Ros2TraceAnalyzer]: https://github.com/skoudmar/Ros2TraceAnalyzer

CC: @wentasah

skoudmar committed:
Add Callback type to allow tracing callback calls

skoudmar committed:
Make rcl_client_t have stable location

This prevents another client initializing at the same location.
Therefore preventing conflicts in a trace.

skoudmar committed:
Make rcl_subscription_t have stable location

This change is necessary because rcl uses address of the rcl handle in
tracepoints. If the location changes further `take` events will not
be correctly assigned to the subscriber.

skoudmar committed:
Make rcl_publisher_t have stable location

This change is necessary because rcl uses address of the rcl handle in
tracepoints. If the location changes, further publish events will not
be correctly assigned to the publisher.

skoudmar committed:
Make rcl_service_t have stable location

This prevents another service initializing at the same location.
Therefore preventing conflicts in a trace.

skoudmar committed:
Add example showing usage of callback tracing

skoudmar committed:
Add tracing to parameter handling services

skoudmar committed:
Add tracepoints to r2r

Location and arguments of tracepoints is consistent with rclcpp

wentasah commented on r2r_tracing/README.md:
```diff
@@ -0,0 +1,46 @@
+# r2r_tracing
+
+Internal dependency containing tracepoint definitions or imports for r2r.
```
Internal dependency containing tracepoint definitions or imports from `tracetools` ROS package for r2r.

wentasah commented on r2r_tracing/README.md:
```diff
@@ -0,0 +1,46 @@
+## Depends on
+
+- `tracetools` ROS package
+  - This package is part of ROS distribution.
```
a part

wentasah commented on r2r_tracing/README.md:
```diff
@@ -0,0 +1,46 @@
+- `lttng-ust` crate
+  - To define additional tracepoints.
+
+## Record traces of R2R applications
```
Recording

wentasah commented on r2r_tracing/README.md:
```diff
@@ -0,0 +1,46 @@
+## Record traces of R2R applications
+
+Make sure to enable feature flag `tracing`
+
```
Then start tracing session:

wentasah commented on r2r_tracing/src/macros.rs:
```diff
@@ -0,0 +1,26 @@
+/// Takes a list of functions and generates two versions of each function:
```
Generates two versions of each function definition wrapped via this macro:

wentasah commented on r2r_tracing/src/rclcpp_tracepoints.rs:
```diff
@@ -0,0 +1,156 @@
+tracepoint_fn! {
+/// Message publication.
+///
+/// Notes the pointer to the `message` being published at the `rclcpp`/`r2r` level.
```
Records

wentasah commented on r2r_tracing/src/rclcpp_tracepoints.rs:
```diff
@@ -0,0 +1,156 @@
+
+/// Subscription object initialisation.
+///
+/// Links the `subscription_handle` from `rcl` to the addresss of rust `subscription` reference.
```
Tracepoint to allow associating ... with ...

wentasah commented on r2r_tracing/src/rclcpp_tracepoints.rs:
```diff
@@ -0,0 +1,156 @@
+    }
+}
+
+/// Link a subscription callback identified by `callback_id` to a `subscription` object.
```
Tracepoint to allow associating...

wentasah commented on r2r_tracing/src/rclcpp_tracepoints.rs:
```diff
@@ -0,0 +1,156 @@
+
+/// Message taking.
+///
+/// Notes the **reference** to the `message` being taken at the `rclcpp`/`r2r` level.
```
Records

wentasah commented on r2r_tracing/src/rclcpp_tracepoints.rs:
```diff
@@ -0,0 +1,156 @@
+///
+/// Notes the **reference** to the `message` being taken at the `rclcpp`/`r2r` level.
+///
+/// To trace void pointer to the message use [`trace_take_ptr`].
```
To trace messages pointed to by void pointer use

wentasah commented on r2r_tracing/src/rclcpp_tracepoints.rs:
```diff
@@ -0,0 +1,156 @@
+    }
+}
+
+/// Link a service callback identified by `callback_id` to a `service`.
```
Tracepoint to allow associtation

wentasah commented on r2r_tracing/src/rclcpp_tracepoints.rs:
```diff
@@ -0,0 +1,156 @@
+    unsafe { tp::ros_trace_rclcpp_service_callback_added(service.cast(), c_void!(callback_id)) }
+}
+
+/// Link a timer callback identified by `callback_id` to its `rcl_timer_t` handle.
```
Tracepoint to allow associtation

wentasah commented on r2r_tracing/src/rclcpp_tracepoints.rs:
```diff
@@ -0,0 +1,156 @@
+    }
+}
+
+/// Link a `timer` to a `node`.
```
Tracepoint to allow associtation

wentasah commented on r2r_tracing/src/rclcpp_tracepoints.rs:
```diff
@@ -0,0 +1,156 @@
+    }
+}
+
+/// Register a demangled `function_symbol` with a `callback_id`.
```
Tracepoint to allow associtation

wentasah commented on r2r_tracing/src/rclcpp_tracepoints.rs:
```diff
@@ -0,0 +1,156 @@
+
+/// Start of a callback
+///
+/// If this callback is done via intra-process set `is_intra_process` to `true`.
```
///  Set `is_intra_process` depending on whether this callback is done via intra-process or not

wentasah commented on r2r_tracing/src/callback.rs:
```diff
@@ -0,0 +1,80 @@
+        }
+    }
+
+    /// Emits trace event attaching this `callback` to the `service`.
```
/// Emits trace event associating this `callback` with the `service`.

wentasah commented on r2r_tracing/src/callback.rs:
```diff
@@ -0,0 +1,80 @@
+        Self::new(callback, id)
+    }
+
+    /// Emits trace event attaching this `callback` to the `timer`.
```
dtto

wentasah commented on r2r_tracing/src/callback.rs:
```diff
@@ -0,0 +1,80 @@
+        Self::new(callback, id)
+    }
+
+    /// Emits trace event attaching this `callback` to the `subscription`.
```
dtto

wentasah commented on r2r/src/time_source.rs:
```diff
@@ -213,6 +216,13 @@ impl TimeSourceSubscriber {
             )?
         };
 
+        // Use TimeSource inner arc address as callback_id. This id should be much larger value than
```
This id should -> We hope that this id is

wentasah commented on r2r/src/nodes.rs:
```diff
@@ -601,6 +601,57 @@ impl Node {
         Ok(receiver)
     }
 
+    /// Subscribe to a ROS topic.
```
Subscribe to a ROS topic and create tracepoints around the callback.

wentasah commented on r2r/src/nodes.rs:
```diff
@@ -601,6 +601,57 @@ impl Node {
+    /// Subscribe to a ROS topic.
+    ///
+    /// This function takes a `callback`, and associates it with the
+    /// subscription. Producing a `Future` representing this task.
```
subscription, producing

wentasah commented on r2r/src/nodes.rs:
```diff
@@ -811,6 +862,61 @@ impl Node {
         Ok(c)
     }
 
+    /// Create a ROS service.
```
... and ...

wentasah commented on r2r/src/nodes.rs:
```diff
@@ -1651,6 +1757,22 @@ impl Timer {
+    ///
+    /// The callback execution is traced by r2r_tracing.
+    ///
+    /// This function should be called immediately after the timer is created.
```
This function should be called before dropping the timer's node. Otherwise, the trace data might be inconsistent.

wentasah commented on r2r/examples/callback_tracing.rs:
```diff
@@ -0,0 +1,70 @@
+/// This example demonstrates creation of a service,
+/// subscriber and timers with their callback execution traced.
+///
+/// Run toghtether with the client example.
+/// e.g. cargo run --example service
+/// and in another terminal cargo run --example client
```
Update

beuss-git committed:
Update bindgen 0.63.0 -> 0.71.1 (#116)

This fixes #116 by now generating the missing primitive sequence functions.
Tested to be working on Ubuntu 22.04 with ROS2 Humble and Iron.

m-dahl committed:
Add `/set_parameters_atomically` (#120, #121)

* Fixed failing `ros2 param ...` on r2r nodes for Jazzy (#120)

Add `/set_parameters_atomically` service to `make_parameter_handler_internal` in `nodes.rs` to fix failing `ros2 param ...` on r2r nodes for Jazzy.

* Atomic behavior for `set_parameters_atomically` (#121).

---------

Co-authored-by: Desmond Germans <desmond@germansmedia.nl>

m-dahl committed:
Release 0.9.5

skoudmar committed:
Create TracingId type

The TracingId is used as a wrapper for the pointer to allow implementing
Send and Sync traits.

The pointer is used only as an ID and is never dereferenced, therefore, it is safe.

skoudmar committed:
Create to Stream wrapper with traced_callback method

The StreamWithTracingData wrapper implements both Stream and FusedStream similarly to Receiver

skoudmar committed:
Add r2r_tracing crate

The crate contains tracepoints; either defined directly or imported
from ros2_tracing project (tracetools package).

The crate define a `tracing` feature which should only be used on Linux
because it uses LTTng only available on Linux.
